---
title: "HUGTiP HIV-1 Drug Resistance Analysis"
author: "Poppy J Hesketh Best"
date: "`r Sys.Date()`"
output: html_document
params:
    sampleID: "insert_sampleID"
---


```{r, echo=FALSE, message=FALSE, warning = FALSE, quiet=TRUE}
library(tidyverse, quiet=TRUE)

# Accessing parameters in R
    sample_ID <- "insert_sampleID"

    read_lengths <- read_csv("read_lenths.tsv")
    colnames(read_lengths) <- c("length")

    read_counts <- read_delim("stats.tsv")

    sierrapy <- read_csv("sierrapy.hiv1.csv")
    drug.groups <- read_csv("drug.groups.csv")

        sierrapy_dr <- sierrapy |> 
            select(-name, -subtype) |>  # Remove name and subtype columns
            pivot_longer(cols = everything(),  # Pivot all remaining columns
                        names_to = "Drug Name", 
                        values_to = "HIVDB Score") |>
            mutate(
            `Drug susceptibility` = case_when(
                `HIVDB Score` >= 0  & `HIVDB Score` <= 9   ~ "Susceptible",
                `HIVDB Score` >= 10 & `HIVDB Score` <= 29  ~ "Low-level resistance",
                `HIVDB Score` >= 30 & `HIVDB Score` <= 59  ~ "Intermediate resistance",
                `HIVDB Score` >= 60                        ~ "High-level resistance",
                TRUE ~ NA_character_  # Handles any unexpected cases
            ))
        sierrapy_dr <- left_join(sierrapy_dr, drug.groups, by = "Drug Name")

    hydra <- read_csv("hydra_report.csv") |>
        mutate(Mutation_code = paste0(Wildtype,Position,Mutation,
                    " (",`Mutation Frequency`,"%)", sep = "")) |>
        mutate(MutationID = paste0(Wildtype,Position,Mutation)
        )

    coverage <- read_delim("coverage_file.csv", col_names = FALSE, delim = "\t")
    colnames(coverage) <- c("chr", "position","coverage")

    hydra_vcf <- read_delim("hydra.vcf", comment = "#", delim = '\t', col_names = FALSE)
    colnames(hydra_vcf) <- c("CHROM","position","GENE","REF","ALT","COVERAGE","FILTER","INFO")


    run_params <- read_csv("run_params.csv", col_names = FALSE)
    colnames(run_params) <- c("param", "val")
    parameters <- run_params |> 
        pivot_wider(names_from = param, values_from = val)

# Create the final mutations files for reporting the results
    mutations_data <- read_delim("quasitools-mutation_db.tsv", delim = "\t") |> 
        mutate(MutationID = paste0(Wildtype,Position,Mutation))

    if (nrow(hydra) > 0) {

        hydra_final <- left_join(hydra, mutations_data)
        hydra_PR <- hydra_final |> filter(Gene == "PR")
        hydra_RT <- hydra_final |> filter(Gene == "RT")
        hydra_IN <- hydra_final |> filter(Gene == "IN")

    } else {

    # Create empty data frames with the desired structure, even when no data in hydra
    empty_columns <- c("Chromosome", "Gene", "Category", "Surveillance", 
                        "Wildtype", "Position", "Mutation", 
                        "Mutation Frequency", "Coverage",
                        "Mutation_code", "MutationID")
    
    hydra_PR <- data.frame(matrix(ncol = length(empty_columns), nrow = 0))
    colnames(hydra_PR) <- empty_columns
    
    hydra_RT <- data.frame(matrix(ncol = length(empty_columns), nrow = 0))
    colnames(hydra_RT) <- empty_columns
    
    hydra_IN <- data.frame(matrix(ncol = length(empty_columns), nrow = 0))
    colnames(hydra_IN) <- empty_columns

    }

```

-------------------------------------------------------------------------------------------

## Sequencing Summary

-------------------------------------------------------------------------------------------

### **Query ID**: `r sample_ID`

**Total of reads**: `r scales::comma(sum(read_counts$num_seqs, na.rm = TRUE))` (PE: `r scales::comma(mean(read_counts$num_seqs, na.rm = TRUE))`)

**HIV subtype**: `r sierrapy$subtype`

-------------------------------------------------------------------------------------------

### Pipeline parameters

**Minimum read depth**: `r parameters$length_cutoff` | **Minimum read quality**: `r parameters$min_read_qual`

**Minimum mutation frequency percent to report**: `r parameters$min_ac`

**The minimum required allele count for variant to be considered**: `r parameters$min_ac`

**Minimum required read depth for variant to be considered**: `r parameters$min_dp`

**The minimum required frequency for mutation to be considered**: `r parameters$min_freq`

**Minimum percentage a base needs to be incorporated into the consensus sequence**: `r parameters$consensus_pct`

**Version control**: 

Quasitools              : `r parameters$quasitools`

Stanford HIVDR Tools/DB : `r parameters$HIVDB`

TrimGalore              : `r parameters$TrimGalore`

SeqKit                  : `r parameters$seqkit`

-------------------------------------------------------------------------------------------

## Drug resistance results 

-------------------------------------------------------------------------------------------

### RNA-dependent polymerase / RdRp

**Total Mutations**: `r if(nrow(hydra_PR) == 0) "None" else nrow(hydra_PR)`

**Major Resistance Mutations (Frequency %)**: `r if(nrow(hydra_PR |> filter(str_detect(Category, "Major" ))) == 0) "None" else hydra_PR |> filter(str_detect(Category, "Major")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Minor Resistance Mutations (Frequency %)**: `r if(nrow(hydra_PR |> filter(str_detect(Category, "Minor|PIMinor" ))) == 0) "None" else hydra_PR |> filter(str_detect(Category, "Minor")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Accessory Resistance Mutations (Frequency %)**: `r if(nrow(hydra_PR |> filter(str_detect(Category, "Accessory"))) == 0) "None" else hydra_PR |> filter(str_detect(Category, "Accessory")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Other Mutations (Frequency %)**: `r if(nrow(hydra_PR |> filter(str_detect(Category, "Other"))) == 0) "None" else hydra_PR |> filter(str_detect(Category, "Other")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

Comments:

`r if (nrow(hydra_PR) == 0) "None" else hydra_PR |> select(MutationID, Comment) |> mutate(combined = paste0(Comment)) |> pull(combined) |> paste(collapse = "\n\n")`

-----

### Reverse Trascriptase  / RT

**Total Mutations**: `r if(nrow(hydra_RT) == 0) "None" else nrow(hydra_RT)`

**Major NNRTI Resistance Mutations (Frequency %)**: `r if(nrow(hydra_RT |> filter(str_detect(Category, "^NNRTI"))) == 0) "None" else hydra_RT |> filter(str_detect(Category, "Major|NNRTI")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Major NRTI Resistance Mutations (Frequency %)**: `r if(nrow(hydra_RT |> filter(str_detect(Category, "^NRTI"))) == 0) "None" else hydra_RT |> filter(str_detect(Category, "Major|NRTI")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Minor Resistance Mutations (Frequency %)**: `r if(nrow(hydra_RT |> filter(str_detect(Category, "Minor"))) == 0) "None" else hydra_RT |> filter(str_detect(Category, "Minor")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Accessory Resistance Mutations (Frequency %)**: `r if(nrow(hydra_RT |> filter(str_detect(Category, "Accessory"))) == 0) "None" else hydra_RT |> filter(str_detect(Category, "Accessory")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Other Mutations (Frequency %)**: `r if(nrow(hydra_RT |> filter(str_detect(Category, "Other"))) == 0) "None" else hydra_RT |> filter(str_detect(Category, "Other")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

Comments:

`r if (nrow(hydra_RT) == 0) "None" else hydra_RT |> select(MutationID, Comment) |> mutate(combined = paste0(Comment)) |> pull(combined) |> paste(collapse = "\n\n")`

-----

### Intergrase / IN

**Total Mutations**: `r if(nrow(hydra_IN) == 0) "None" else nrow(hydra_IN)`

**Major Resistance Mutations (Frequency %)**: `r if(nrow(hydra_IN |> filter(str_detect(Category, "Major"))) == 0) "None" else hydra_IN |> filter(str_detect(Category, "Major")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Minor Resistance Mutations (Frequency %)**: `r if(nrow(hydra_IN |> filter(str_detect(Category, "Minor"))) == 0) "None" else hydra_IN |> filter(str_detect(Category, "Minor")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Accessory Resistance Mutations (Frequency %)**: `r if(nrow(hydra_IN |> filter(str_detect(Category, "Accessory"))) == 0) "None" else hydra_IN |> filter(str_detect(Category, "Accessory")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

**Other Mutations (Frequency %)**: `r if(nrow(hydra_IN |> filter(str_detect(Category, "Other"))) == 0) "None" else hydra_IN |> filter(str_detect(Category, "Other")) |> select(Mutation_code) |> deframe() |> paste(collapse = ", ")`

Comments:

`r if (nrow(hydra_IN) == 0) "None" else hydra_IN |> select(MutationID, Comment) |> mutate(combined = paste0(Comment)) |> pull(combined) |> paste(collapse = "\n\n")`

-------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------

## Drug Resistance HIVDB Scores

```{r}
#| label: fig-HIVDB_SCORES
#| fig-cap: "HIVDB Drug Resistance Scores"
#| warning: false

library(gt)

# Format and print table using gt
sierrapy_dr |>
    arrange(`HIVDB Score`, `Target`)  |>
    select(`Target`, `Drug Name`, `HIVDB Score`, `Drug susceptibility`) |>
    gt()
``` 

-------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------

## Sequencing analysis summary

### Read fragments

```{r fig-read-histogram, fig.cap="Histogram of read lengths", warning=FALSE}
#| label: fig-read-histogram
#| fig-cap: "Histogram of read lengths"
#| warning: false

library(tidyverse); library(plotly)

p1 <- read_lengths |>
    mutate(sample = paste(sample_ID)) |>
    ggplot(aes(x=length)) +
        geom_histogram(binwidth=5, fill = "darkmagenta") +
    ggtitle("Binwidth = 5") +
    labs(x = "Read length (bp)", y = "Frequency") +
    theme_bw()

ggplotly(p1)

```

### Genome coverage

```{r fig-read-histogram, fig.cap="HXB2 Polymerase coverage of read lengths", warning=FALSE}
#| label: fig-hiv_coverage
#| fig-cap: "Annotated genome with read coverage."
#| warning: false

library(tidyverse)

# Generate full sequence of positions
full_positions <- tibble(position = full_seq(coverage$position, 1))

# Join with existing data, filling missing positions with NA
coverage_complete <- full_positions %>%
    left_join(coverage, by = "position")

# create the plot
p2 <- ggplot(coverage_complete, 
            aes(x = position, y = coverage)
            ) + 
    geom_line(na.rm = TRUE) +
    theme_bw() +
    theme(
        legend.position = "none",
        plot.title = element_text(size = 11)) +
    labs(x = "Position along HXB2 pol", y = "Depth of coverage (no. reads)") +
    geom_rect(aes(xmin = 0, xmax = 296, ymin = -200, ymax = 0), fill = "blue", alpha = 0.3) +
    annotate("text", x = 148, y = -900, label = "PR") + 
    geom_rect(aes(xmin = 297, xmax = 1616, ymin = -500, ymax = -300), fill = "red", alpha = 0.3) +
    annotate("text", x = 956, y = -900, label = "RT") + 
    geom_rect(aes(xmin = 1977, xmax = 2843, ymin = -200, ymax = 0), fill = "green", alpha = 0.3) +
    annotate("text", x = 2410, y = -900, label = "IN") 

ggplotly(p2)

```

-------------------------------------------------------------------------------------------

## Methods

Illumina reads are quality controlled and trimmed of known Illumina adaptors with Trim_galore, fastQC/multiQC are used to visualise the overall quality of the Illumina reads. Trimmed reads are parsed to Quasitools HYDRA that performs read mapping to the HIV-1 HxB2 reference genomes using bowtie2. Variants are called and queried against a list of known mutations from the Stanford HIV resistance mutation database. A consensus genomes is also generated. The conensus genome is then submitted to the Stanford HIVDB server and drug resistance scores are calcualted (@fig-HIVDB_SCORES). Seqkit and Samtools are used to generated covarage profiles across the genomes (@fig-hiv_coverage) and read length frequencies (@fig-read-histogram).

-------------------------------------------------------------------------------------------

## Citations

Ewels P. et al.  (2016) [MultiQC: summarize analysis results for multiple tools and samples in a single report.](https://doi.org/10.1093/bioinformatics/btw354) Bioinformatics, 32, 3047–3048.

Ho J.C. et al.  (2019) [Sierra-local: a lightweight standalone application for drug resistance prediction.](https://joss.theoj.org/papers/10.21105/joss.01186.pdf) Softw. J. Open Source Softw., 4, 1186.

Krueger F. (2012) [Trim Galore: A Wrapper Tool Around Cutadapt and FastQC to Consistently Apply Quality and Adapter Trimming to FastQ files, with Some Extra Functionality for MspI-Digested RRBS-Type (Reduced Representation Bisufite-Seq) Libraries.](http://www.bioinformatics.babraham.ac.uk/projects/trim\_galore/)

Langmead B., Salzberg S.L. (2012) [Fast gapped-read alignment with Bowtie 2.](https://www.nature.com/articles/nmeth.1923) Nat. Methods, 9, 357–359.
